lib/base_service.js

```diff
@@ -31,13 +32,48 @@ class BaseService {
         this._logger = null;
         this._metrics = null;
         this._ratelimiter = null;
-        this.serviceReturns = null;
+        this._serviceReturns = null;
+        this._uncaughtExceptionHandlerAdded = false;
+        this._unhandledRejectionHandlerAdded = false;
+
+        if (!this._uncaughtExceptionHandlerAdded) {
+            this._uncaughtExceptionHandlerAdded = true;
+            this._uncaughtExceptionHandler = (err) => {
+                this._logUnhandledException(err);
+                // Don't exit right away, allow logger to process message
+                setTimeout(() => {
+                    process.exit(1);
+                }, 100);
+            };
+        }
+        // Avoid recursion if there are bugs in the logging code.
+        this.inLogger = false;
+
+        if (!this._unhandledRejectionHandlerAdded) {
+            this._unhandledRejectionHandlerAdded = true;
+            this._logUnhandledException = (err) => {
+                if (!this.inLogger) {
+                    this.inLogger = true;
+                    this.log('fatal/service-runner/unhandled', err);
+                    this.inLogger = false;
+                }
+            };
+        }
     }
 
     stop() {
         if (this._ratelimiter) {
             this._ratelimiter.stop();
         }
+        // Remove handlers
+        if (this._logger && this._logger.removeListener) {
+            if (this._logger.removeListener) {
+                this._logger.removeListener('error', this._logger.errorHandler);
+            }
+
+            process.removeListener('unhandledRejection', this._logUnhandledException);
+            process.removeListener('uncaughtException', this._uncaughtExceptionHandler);
+        }
```

Comment:
This doesn't reset the `unhandled*Added booleans, so whe we restart it back the new handlers will not be re-added. 

Since we don't need the booleans, we need to check if the handlers are not undefined, remove them and check whether the candlers are not undefined.

Reply:
Getting rid of the booleans

Reply:
Done